components/match2/wikis/callofduty/match_group_input_custom.lua

```diff
@@ -0,0 +1,491 @@
+			end
+			opponents[opponentIndex] = opponent
+
+			-- get players from vars for teams
+			if opponent.type == Opponent.team and not Logic.isEmpty(opponent.name) then
+				match = matchFunctions.getTeamPlayers(match, opponentIndex, opponent.name)
+			end
+		end
+	end
+
+	-- see if match should actually be finished if bestof limit was reached
+	if isScoreSet and not Logic.readBool(match.finished) then
+		local firstTo = math.ceil(match.bestof / 2)
+		for _, item in pairs(opponents) do
+			if (tonumber(item.score or 0) or 0) >= firstTo then
+				match.finished = true
+				break
+			end
+		end
+	end
+
+	-- check if match should actually be finished due to a non score status
+	if not Logic.readBool(match.finished) then
+		for _, opponent in pairs(opponents) do
+			if String.isNotEmpty(opponent.status) and opponent.status ~= 'S' then
+				match.finished = true
+				break
+			end
+		end
+	end
+
+	-- see if match should actually be finished if score is set
+	if isScoreSet and not Logic.readBool(match.finished) and match.hasDate then
+		local currentUnixTime = os.time(os.date('!*t'))
+		local lang = mw.getContentLanguage()
+		local matchUnixTime = tonumber(lang:formatDate('U', match.date))
+		local threshold = match.dateexact and SECONDS_UNTIL_FINISHED_EXACT
+			or SECONDS_UNTIL_FINISHED_NOT_EXACT
+		if matchUnixTime + threshold < currentUnixTime then
+			match.finished = true
+		end
+	end
+
+	-- apply placements and winner if finshed
+	if not String.isEmpty(match.winner) or Logic.readBool(match.finished) then
+		match, opponents = CustomMatchGroupInput.getResultTypeAndWinner(match, opponents)
+	end
+
+	-- Update all opponents with new values
+	for opponentIndex, opponent in pairs(opponents) do
+		match['opponent' .. opponentIndex] = opponent
+	end
+	return match
+end
+
+-- Get Playerdata from Vars (get's set in TeamCards) for team opponents
+function matchFunctions.getTeamPlayers(match, opponentIndex, teamName)
+	-- let's make sure we don't leave any gaps.
+	match['opponent' .. opponentIndex].match2players = {}
+	for playerIndex = 1, MAX_NUM_PLAYERS do
+		-- parse player
+		local player = Json.parseIfString(match['opponent' .. opponentIndex .. '_p' .. playerIndex]) or {}
+		player.name = player.name or Variables.varDefault(teamName .. '_p' .. playerIndex)
+		player.flag = player.flag or Variables.varDefault(teamName .. '_p' .. playerIndex .. 'flag')
+		player.displayname = player.displayname or Variables.varDefault(teamName .. '_p' .. playerIndex .. 'dn')
+		if not Table.isEmpty(player) then
+			table.insert(match['opponent' .. opponentIndex].match2players, player)
+		end
+	end
+
+	return match
+end
+
+-- Get Playerdata for solo opponents
+function matchFunctions.readSoloOpponent(match, opponentIndex)
+	local player = Json.parseIfString(match['opponent' .. opponentIndex .. '_p' .. playerIndex]) or {}
+	player.name = player.name or 'TBD'
+	player.flag = player.flag
+	player.displayname = player.displayname or player.name
+	return {player}
```

Comment:
possibly you might want the input to be directly in the opponent instead of inside an extra player struct inside the player
maybe ask what the wiki wants

if inside opponent then do something like
```suggestion
	local name = oppoent.name or opponent[1] or 'TBD'
	
	return = {{
		name = name,
		flag = opponent.flag,
		displayname = opponent.displayname or name,
	}}
```

Reply:
> possibly you might want the input to be directly in the opponent instead of inside an extra player struct inside the player maybe ask what the wiki wants
> 
> if inside opponent then do something like

Are there significant changes on the input, if there isnt any (nothing changes on how to input stuff) I think we can go with that idea right away.

(but currently the code when pasted it will return an unexpected symbol)

Reply:
well currently you would have to do something like
`|opponent1={{SoloOpponent|name=}}|opponent1_p1={{BrktsPlayer|name=|displayname=|flag=}}`
with the suggestion you can just use
`|opponent1={{SoloOpponent|name=|displayname=|flag=}}`

Reply:
Noted, the suggestion is obviously the ultimate go-to choice
![image](https://user-images.githubusercontent.com/88981446/205860261-06a7c86e-1d24-44bf-b691-591f95ad6f86.png)
 
I've tried pasted and it returns an unexpected symbol on 429 here. Not sure what I have missed out

Reply:
> local name = oppoent.name or opponent[1] or 'TBD'
> 	
> 	return = {{
> 		name = name,
> 		flag = opponent.flag,
> 		displayname = opponent.displayname or name,
> 	}}

Ok it is now able to be saved, Before I can update the revision onto this PR theres one last thing to do is to make solo no longer an **unsupported type**: Dev at https://liquipedia.net/callofduty/User:Hesketh2/Test#Solo_Test (has the suggestion above already added)

https://liquipedia.net/callofduty/index.php?title=Module:MatchGroup/Input/Custom/dev&action=edit

If I removed 83-86 Solo will work but team will throw error instead

After this getting fixed I think it should be all good now

Reply:
https://liquipedia.net/callofduty/index.php?title=Module%3AMatchGroup%2FInput%2FCustom%2Fdev&type=revision&diff=108059&oldid=108052